controller/deployer/strategies/all-at-once.go

```diff
@@ -6,67 +6,67 @@ func allAtOnce(d *Deploy) error {
 	log := d.logger.New("fn", "allAtOnce")
 	log.Info("starting all-at-once deployment")
 
-	olog := log.New("release_id", d.OldReleaseID)
-	olog.Info("getting old formation")
-	f, err := d.client.GetFormation(d.AppID, d.OldReleaseID)
-	if err != nil {
-		olog.Error("error getting old formation", "err", err)
-		return err
-	}
-
-	nlog := log.New("release_id", d.NewReleaseID)
-	nlog.Info("creating new formation", "processes", f.Processes)
-	if err := d.client.PutFormation(&ct.Formation{
-		AppID:     d.AppID,
-		ReleaseID: d.NewReleaseID,
-		Processes: f.Processes,
-	}); err != nil {
-		nlog.Error("error creating new formation", "err", err)
-		return err
-	}
-
 	expected := make(jobEvents)
-	for typ, n := range f.Processes {
-		for i := 0; i < n; i++ {
+	for typ, n := range d.Processes {
+		existing := d.newReleaseState[typ]
+		for i := existing; i < n; i++ {
 			d.deployEvents <- ct.DeploymentEvent{
 				ReleaseID: d.NewReleaseID,
 				JobState:  "starting",
 				JobType:   typ,
 			}
 		}
-		expected[typ] = map[string]int{"up": n}
-	}
-	nlog.Info("waiting for job events", "expected", expected)
-	if err := d.waitForJobEvents(d.NewReleaseID, expected, nlog); err != nil {
-		nlog.Error("error waiting for job events", "err", err)
-		return err
+		expected[typ] = map[string]int{"up": n - existing}
 	}
+	if expected.Count() > 0 {
+		nlog := log.New("release_id", d.NewReleaseID)
+		nlog.Info("creating new formation", "processes", d.Processes)
+		if err := d.client.PutFormation(&ct.Formation{
+			AppID:     d.AppID,
+			ReleaseID: d.NewReleaseID,
+			Processes: d.Processes,
+		}); err != nil {
+			nlog.Error("error creating new formation", "err", err)
+			return err
+		}
 
-	olog.Info("scaling old formation to zero")
-	if err := d.client.PutFormation(&ct.Formation{
-		AppID:     d.AppID,
-		ReleaseID: d.OldReleaseID,
-	}); err != nil {
-		log.Error("error scaling old formation to zero", "err", err)
-		return err
+		nlog.Info("waiting for job events", "expected", expected)
+		if err := d.waitForJobEvents(d.NewReleaseID, expected, nlog); err != nil {
+			nlog.Error("error waiting for job events", "err", err)
+			return err
+		}
 	}
 
 	expected = make(jobEvents)
-	for typ, n := range f.Processes {
-		for i := 0; i < n; i++ {
+	for typ := range d.Processes {
+		existing := d.oldReleaseState[typ]
+		for i := 0; i < existing; i++ {
 			d.deployEvents <- ct.DeploymentEvent{
 				ReleaseID: d.OldReleaseID,
 				JobState:  "stopping",
 				JobType:   typ,
 			}
 		}
-		expected[typ] = map[string]int{"down": n}
+		expected[typ] = map[string]int{"down": existing}
 	}
-	olog.Info("waiting for job events", "expected", expected)
-	if err := d.waitForJobEvents(d.OldReleaseID, expected, olog); err != nil {
-		olog.Error("error waiting for job events", "err", err)
-		return err
+	if expected.Count() > 0 {
+		olog := log.New("release_id", d.OldReleaseID)
```

Comment:
I'd just shadow `log` here and above (to avoid mistakes like the one made on line 59.